rir/src/interpreter/interp.c

```diff
@@ -333,21 +345,56 @@ static SEXP rirCallClosure(SEXP call, SEXP env, SEXP callee, SEXP actuals,
             optimizing = true;
 
             Function* oldFun = fun;
-            SEXP oldBody = body;
-            cp_pool_add(ctx, oldBody);
-            body = globalContext()->optimizer(callee);
+            SEXP oldFunStore = funStore;
+            cp_pool_add(ctx, oldFunStore);
+            funStore = globalContext()->optimizer(funStore);
 
             // TODO: there might be promises with references to the old code!
             // Therefore we keep it around.
             // TODO: first I tried to use R_PreserveObject, but it did not work
             // for large vectors, not sure why, need to investigate
-            cp_pool_add(ctx, body);
+            cp_pool_add(ctx, funStore);
+
+            oldFun->next = funStore;
+            fun = sexp2function(funStore);
+            fun->origin = oldFunStore;
+
+            // Add the new function to the vtable.
+            if (vtable->length < TRUELENGTH(vtableStore)) {
```

Comment:
This is the trickiest bit of code, because we have to re-allocate the table, copy pointers over, and update pointers.

I *think* it's fine, but I've stared at this for so long that I'm no longer sure. So more eyeballs here would help.

Because the tests don't really cover the case (only one optimization is done, so the table of size 1 grows to a table of size 2 and that's it), I hacked the code to compile a new function version *every* time a closure is called. This passed the RIR tests, and I tried running the GNU R tests. It ate up all my RAM, so I killed it, but it didn't segfault.